fs/dafsserver/dafsserver.cpp

```diff
@@ -1288,29 +1295,86 @@ class CRemoteDiskReadActivity : public CRemoteDiskBaseActivity
             return true;
     }
 public:
-    CRemoteDiskReadActivity(IPropertyTree &config) : PARENT(config), prefetchBuffer(nullptr)
+    CRemoteStreamReadBaseActivity(IPropertyTree &config, IFileDescriptor *fileDesc) : PARENT(config, fileDesc)
     {
         compressed = config.getPropBool("compressed");
+        chooseN = config.getPropInt64("chooseN", defaultFileStreamChooseNLimit);
+
+        partNum = config.getPropInt("virtualFields/partNum");
+        baseFpos = (offset_t)config.getPropInt64("virtualFields/baseFpos");
+    }
+    ~CRemoteStreamReadBaseActivity()
+    {
+        delete filterRow;
+    }
+    virtual void init(IPropertyTree &config) override
+    {
+        PARENT::init(config);
+        if (config.hasProp("keyFilter"))
+        {
+            filterRow = new RtlDynRow(*record);
+            Owned<IPropertyTreeIterator> filterIter = config.getElements("keyFilter");
+            ForEach(*filterIter)
+                filters.addFilter(*record, filterIter->query().queryProp(nullptr));
+        }
+    }
+//interface IVirtualFieldCallback
+    virtual unsigned __int64 getFilePosition(const void * row) override
+    {
+        return inputStream->tell() + baseFpos;
+    }
+    virtual unsigned __int64 getLocalFilePosition(const void * row) override
+    {
+        return makeLocalFposOffset(partNum, inputStream->tell());
+    }
+};
+
+class CRemoteDiskReadActivity : public CRemoteStreamReadBaseActivity
+{
+    typedef CRemoteStreamReadBaseActivity PARENT;
+
+    CThorContiguousRowBuffer prefetchBuffer;
+    Owned<ISourceRowPrefetcher> prefetcher;
+    bool inputGrouped = false;
+    mutable bool eogPending = false;
+    mutable bool someInGroup = false;
+    Owned<const IDynamicTransform> translator;
+
+    virtual bool refreshCursor() override
+    {
+        if (prefetchBuffer.tell() != startPos)
```

Comment:
it can, but needs to check CThorContiguousRowBuffer::tell() ..
So could make it:
```
    virtual bool refreshCursor() override
    {
        if (prefetchBuffer.tell() != startPos)
        {
            PARENT::refreshCursor();
            prefetchBuffer.clearStream();
            prefetchBuffer.setStream(inputStream);
            return true;
        }
        return false;
    }

```